### PR TITLE
(1047) Document our approach to building the caselist API and caching

### PR DESCRIPTION
**Pull request by Gweaton**
## Context

https://trello.com/c/eKCVfadw/1047-spike-how-do-we-handle-fetching-and-caching-data-from-other-services-for-the-caselist-endpoint-m

## Changes in this PR

Adds a couple of ADRs to document our approaches to building the API, including how we'll handle caching if we decide to go down that route.

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes UI for this change to work...
  - [ ] ... and they been released to production already

### Post-merge

- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

**Review by nsmj63-dxw — APPROVED** [review]

File: doc/adr/0006-avoid-caching-caselist-endpoint-data-until-we-re-sure-we-need-it.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+lots of requests being made to other services and slow our service down.
+
+## Decision
+
+Because we're making a request to three external services here for potentially a
+small amount of data, we're going to delay implementing caching until we know a
+bit more about how the service will perform, as we're hesistant to overengineer
+an approach at this stage.
+
+We will monitor the performance of our caselist endpoint and implement caching
```
Comment (anandelangovan1) [inline doc/adr/0006-avoid-caching-caselist-endpoint-data-until-we-re-sure-we-need-it.md:26]: Do we have dashboards/metrics to monitor our application?

File: doc/adr/0006-avoid-caching-caselist-endpoint-data-until-we-re-sure-we-need-it.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+## How we plan to implement caching if it becomes necessary
+
+This approach has been largely inspired by the [CAS team's
+implementation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4545904995/API+caching),
+which involves two different types of caching:
+
+### Pre-emptive cache
+
+This runs on startup and renews itself every 6 hours (an API trick, not done via
```
Comment (anandelangovan1) [inline doc/adr/0006-avoid-caching-caselist-endpoint-data-until-we-re-sure-we-need-it.md:51]: Don't quite understand this. Are we saying we are going to cache prisonNumber separately and if we were to call person api then cache person details separately?

Comment (Gweaton) [reply doc/adr/0006-avoid-caching-caselist-endpoint-data-until-we-re-sure-we-need-it.md]: Have rephrased this section to be a bit clearer.

File: doc/adr/0006-avoid-caching-caselist-endpoint-data-until-we-re-sure-we-need-it.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+CI or a cron job) and expires after 12 hours of no response; this is
+opinionated, e.g. for our service we'd specifically look out for a request for
+any Prison Number, which may be a candidate for going to various external
+endpoints and caching the response - we'd be displaying user details like
+Sentence Dates, Name etc to multiple users viewing their caselists, so making a
+request each time would be inefficient.
+
+### A regular cache
+
+This handles traditionally and does not renew itself; is not opinionated, if a
```
[the remark this answers is not in the document]
Comment (Gweaton) [reply doc/adr/0006-avoid-caching-caselist-endpoint-data-until-we-re-sure-we-need-it.md:65]: Good point, I've just added a line to clarify we'll want to add this functionality.

**Pull request merged**